### ISSUE TITLE
Safely generate display name for @ParameterizedTest if argument.toString() throws exception

### ISSUE THREAD

**Issue by johnsoneal** [junit-team/junit5#1492]
## Overview

Using JUnit 5.2.0

I am using parameterized tests, and for the most part everything works as expected. However the name with pattern fails if any object has a bad `toString()`, even if the object is not referenced by the pattern. For example:

```java
public class BadToStringTest {

    static Object[] data() {
        return new Object[][] { { "has a bad toString()", new BadToString() } };
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("data")
    void test(String name, Object bad) {
        // Empty Block
    }

    static class BadToString {
        @Override
        public String toString() {
            throw new RuntimeException("Very bad to string");
        }
    }
}
```

In the example an exception is thrown as all arguments are formatted not just the specified pattern.

I understand that throwing an exception from `toString()` is a really bad idea; however, it is somewhat unexpected that the formatter is calling `toString()` on arguments that it is not explicitly being asked to format.

As a side note, Hamcrest for a similar situation uses a `toString()` that conforms to `Object#toString()`:

```java
try {
    return String.valueOf(value);
}
catch (Exception e) {
    return value.getClass().getName() + "@" + Integer.toHexString(value.hashCode());
}
```

## Deliverables

- [x] Ensure that invocation of `toString()` on arguments supplied to a `@ParameterizedTest` method does not prevent generation of the display name.
  - In other words, invoke `toString()` _safely_ (within a try-catch block).

**Comment by sbrannen** [comment]
Thanks for raising the issue and including a test case that reproduces the behavior!

**Comment by sbrannen** [comment]
### Analysis

This issue occurs because `toString()` is invoked on all "consumed arguments" in `ParameterizedTestNameFormatter.makeReadable(Object[])`.

In the provided example `Object bad` is _consumed_ by the method even though it is not included in the custom display name.

**Comment by sbrannen** [comment]
FYI: I added _Deliverables_ and slated this for 5.3 RC1.

**Comment by marcphilipp** [comment]
If/when we implement this, it should be a common utility that we can reuse for assertions etc. as well.

**Comment by sbrannen** [comment]
I also updated the title accordingly.

**Comment by sbrannen** [comment]
@marcphilipp, the code in question already uses `org.junit.platform.commons.util.StringUtils.nullSafeToString(Object)`.

So we could either add the exception handling support there or introduce a new method (e.g., `safeToString(Object)`) that delegates to `nullSafeToString()` but returns a default _toString()_ if an exception occurs.

**Comment by sbrannen** [comment]
FWIW, `org.junit.jupiter.api.AssertionUtils.toString(Object)` also currently delegates to `StringUtils.nullSafeToString(Object)`.

**Comment by marcphilipp** [comment]
I think changing `nullSafeToString()` is fine. Having both `safeToString()` and `nullSafeToString()` would be weird IMHO.

In any case, I think we shouldn't call `toString()` unnecessarily for `@ParameterizedTest` arguments.

**Comment by sbrannen** [comment]
@johnsoneal, out of curiosity, how did you realize you could return a 2D array from a `@MethodSource` factory method?

I don't think I've ever seen anyone do that, and the JavaDoc explicitly states the following.

> If a parameterized test method declares multiple parameters, factory methods must return instances of `Arguments`, e.g. as `Stream<Arguments>`.

So, I'm guessing we should update the JavaDoc for `@MethodSource`. 😉

**Comment by marcphilipp** [comment]
... and add a test case for it. 😉

**Comment by sbrannen** [comment]
`@MethodSource` factory methods can return anything supported by `org.junit.platform.commons.util.CollectionUtils.toStream(Object)`. So we need to update the JavaDoc for `CollectionUtils.toStream(Object)` and then sync that with the class-level JavaDoc for `@MethodSource`.

**Comment by johnsoneal** [comment]
@sbrannen 

> out of curiosity, how did you realize you could return a 2D array

Nice to see I have done something not seen by you guys before! 

Their where two main reason for doing it this way. First the `ParameterizedTest` documentation references every thing as being indexed in a `ArgumentsProvider`:

> In this context, an indexed argument is an argument for a given index in the Arguments provided by an ArgumentsProvider that is passed as an argument to the parameterized method at the same index in the method's formal parameter list. 

And secondly the documentation for `ObjectArrayArguments`

> It provides access to an array of objects to be used for invoking a @ParameterizedTest method. 

To me this implied I could supply a 2D array and the _automatic type conversion_ would handle the creation of the `ObjectArrayArguments`.

This was my reasoning, however I guess from your comment this is not the intended usage.

**Comment by sbrannen** [comment]
> Nice to see I have done something not seen by you guys before!

I agree: quite amusing.

`ObjectArrayArguments`? Where did you come across that? I think that's be gone for half a year or more. That's just `Arguments` now. 😉 

> To me this implied I could supply a 2D array and the automatic type conversion would handle the creation of the `Arguments`.

We ultimately use a `Stream<Arguments>` internally. So the automatic type conversion support doesn't come into play here. Rather, it's the simple fact that `Object[][]` is and `instanceof` `Object[]`, and our `CollectionUtils.toStream(Object)` method supports converting an `Object[]` into a `Stream`. So in this particular case, with a 2D object array, we get back a `Stream<Object[]>` and then an additional internal stream converts each `Object[]` into an `Arguments` object, and we ultimately have a `Stream<Arguments>`. 😈

**Comment by sbrannen** [comment]
> This was my reasoning, however I guess from your comment this is not the intended usage.

True... it wasn't intentional. But... if it works, it works.

**Comment by sbrannen** [comment]
I just realized this is the _Columbus_ issue: "In August #1492, Columbus sailed the ocean blue."

**Comment by jbduncan** [comment]
Wow, JUnit 5 has now received almost 1500 issues and PRs! I'd consider that a great milestone. :clinking_glasses:

**Comment by sbrannen** [comment]
**Update**:

> So, I'm guessing we should update the JavaDoc for `@MethodSource`.

✅ 

> ... and add a test case for it.

✅ 

> So we need to update the JavaDoc for `CollectionUtils.toStream(Object)` and then sync that with the class-level JavaDoc for `@MethodSource`.

✅

**Comment by sbrannen** [comment]
Addressed in 5bada91048eb1e9bcc6fba2d37cdee8f627e6f6a.
